File: scripts/optimize_all_stations.py
import pandas as pd
import numpy as np
from scipy.optimize import minimize
import os
import json

# 聚类时段映射
hour_labels = {
    'Lifestyle Services': {1: [10, 11, 14, 15, 16, 17, 18], 2: [7, 8, 9, 12, 13, 19, 20, 21], 3: [22, 23],
                           4: [0, 1, 2, 3, 4, 5, 6]},
    'Food and Beverage Services': {1: [14, 15, 16, 17, 18, 19], 2: [8, 9, 10, 20, 21, 22, 23],
                                   3: [2, 3, 4, 5, 6, 7, 11, 12, 13], 4: [0, 1]},
    'Business and Residential': {1: [10, 11, 14, 15, 16, 17, 18], 2: [7, 8, 9, 12, 13, 19, 20], 3: [21, 22, 23],
                                 4: [0, 1, 2, 3, 4, 5, 6]}
}

# 论文参数（每个类别使用相同的参数）
categories_params = {
    'Lifestyle Services': {'rho': -0.0724, 'beta1': -5.5852, 'beta2': 0.4636, 'const': 6.4749, 'b_t': 0.0038},
    'Business and Residential': {'rho': -0.2358, 'beta1': -2.2836, 'beta2': 1.5628, 'const': 5.0514, 'b_t': 0.0045},
    'Food and Beverage Services': {'rho': 0.0274, 'beta1': -6.0837, 'beta2': -0.2428, 'const': 6.6915, 'b_t': 0.0052}
}

T, S_MIN, S_MAX, PHI_MAX = 24, 0.20, 0.80, 50
C_BASE, ETA = 0.05, 0.2

# TAZ类别映射（基于poiCategory）
TAZ_CATEGORY_MAP = {
    'lifestyle services': 'Lifestyle Services',
    'food and beverage services': 'Food and Beverage Services',
    'business and residential': 'Business and Residential'
}


def run_optimization_all():
    path = "./data/"
    inf_df = pd.read_csv(os.path.join(path, 'inf.csv'))
    e_price_df = pd.read_csv(os.path.join(path, 'e_price.csv')).iloc[:T]
    s_price_hist_df = pd.read_csv(os.path.join(path, 's_price.csv')).iloc[:T]
    occ_df = pd.read_csv(os.path.join(path, 'occupancy.csv')).iloc[:T]
    dist_mat = np.loadtxt(os.path.join(path, 'distance.csv'), delimiter=',')
    adj_mat = np.loadtxt(os.path.join(path, 'adj.csv'), delimiter=',')

    # 获取所有唯一的TAZ
    unique_tazs = []
    for tid in inf_df['TAZID']:
        if tid not in unique_tazs:
            unique_tazs.append(tid)
        if len(unique_tazs) == 276:
            break
    taz_to_idx = {taz: i for i, taz in enumerate(unique_tazs)}
    ts_taz_ids = [int(col) for col in e_price_df.columns if col.isdigit()]

    # 获取每个TAZ的类别（基于站点数据）
    stations_data_path = os.path.join(path, '../data/stations-data.js')
    # 读取站点数据获取poiCategory
    import re
    with open(os.path.join(path, '../data/stations-data.js'), 'r') as f:
        content = f.read()
    # 解析TAZ_POINTS
    taz_cat_map = {}  # tazid -> category
    matches = re.findall(r'\{"tazid":\s*"(\d+)",[^}]*poiCategory":\s*"([^"]+)"', content)
    for tazid, poi_cat in matches:
        cat = TAZ_CATEGORY_MAP.get(poi_cat.lower(), 'Business and Residential')
        taz_cat_map[int(tazid)] = cat

    print(f"Total TAZs with category: {len(taz_cat_map)}")

    # 结果存储
    all_results = {}

    # 按类别分组处理
    for cat_name, params in categories_params.items():
        print(f"\n{'='*60}")
        print(f"Processing: {cat_name}")
        print(f"{'='*60}")

        # 获取该类别的所有站点
        cat_tazs = [taz for taz in unique_tazs if taz in taz_cat_map and taz_cat_map[taz] == cat_name]
        cat_tazs = [t for t in cat_tazs if t in ts_taz_ids and t in taz_to_idx]
        N = len(cat_tazs)
        print(f"Stations in category: {N}")

        if N == 0:
            print("No stations found!")
            continue

        # 提取数据
        taz_counts = inf_df[inf_df['TAZID'].isin(cat_tazs)].groupby('TAZID')['charge_count'].sum().reindex(cat_tazs).values
        E = e_price_df[[str(tid) for tid in cat_tazs]].values
        S_h = s_price_hist_df[[str(tid) for tid in cat_tazs]].values
        Occ = occ_df[[str(tid) for tid in cat_tazs]].values

        # 空间矩阵
        idx = [taz_to_idx[tid] for tid in cat_tazs]
        W = adj_mat[np.ix_(idx, idx)] * (1.0 / (dist_mat[np.ix_(idx, idx)] ** 2 + 1e-6))
        np.fill_diagonal(W, 0)
        row_sums = W.sum(axis=1)
        W = np.divide(W, row_sums[:, np.newaxis], out=np.zeros_like(W), where=row_sums[:, np.newaxis] != 0)
        S_inv = np.linalg.inv(np.eye(N) - params['rho'] * W)

        def calc_profit(S_matrix):
            p_total = 0
            for t in range(24):
                P = np.clip(E[t, :] + S_matrix[t, :], 0.01, 10.0)
                ln_Q = S_inv @ (params['const'] + params['b_t'] * (t + 1) +
                                 params['beta1'] * np.log(P) +
                                 params['beta2'] * (W @ np.log(P)))
                Q_t = np.exp(np.clip(ln_Q, -10, 10))
                cv_t = C_BASE * (1 + ETA * (Occ[t, :] / (Occ[t, :].max() + 1e-6)) ** 2)
                p_total += np.sum(Q_t * (S_matrix[t, :] - cv_t))
            return p_total

        base_p = calc_profit(S_h)
        labels = hour_labels[cat_name]

        def objective(s_flat):
            S_lab = s_flat.reshape((4, N))
            S_full = np.zeros((24, N))
            for l, hs in labels.items():
                S_full[hs, :] = S_lab[l - 1, :]
            return -calc_profit(S_full) / 1000.0

        # 约束
        cons = []
        for t in range(24):
            def cap_con(s_flat, t=t):
                S_lab = s_flat.reshape((4, N))
                curr_l = [l for l, hs in labels.items() if t in hs][0]
                P = np.clip(E[t, :] + S_lab[curr_l - 1, :], 0.01, 10.0)
                ln_Q = S_inv @ (params['const'] + params['b_t'] * (t + 1) +
                                 params['beta1'] * np.log(P) +
                                 params['beta2'] * (W @ np.log(P)))
                return (taz_counts * PHI_MAX) - np.exp(np.clip(ln_Q, -10, 10))
            cons.append({'type': 'ineq', 'fun': cap_con})

        print(f"Optimizing {N} stations with {4*N} variables...")
        res = minimize(objective, np.full(4 * N, 0.5), method='SLSQP',
                       bounds=[(S_MIN, S_MAX)] * (4 * N),
                       constraints=cons,
                       options={'maxiter': 2000, 'ftol': 1e-4})

        print(f"Optimization result: {res.message}")

        if res.success or 'Iteration limit' in res.message:
            opt_p = -res.fun * 1000.0
            S_opt_lab = res.x.reshape((4, N))

            print(f"\nBase Profit: {base_p:.2f}, Opt Profit: {opt_p:.2f}")
            print(f"Profit Increase: {(opt_p - base_p) / abs(base_p) * 100:.2f}%")

            # 存储结果
            cat_results = {
                'profitIncrease': (opt_p - base_p) / abs(base_p) * 100,
                'occIncrease': 0,  # 计算占用率变化
                'baseProfit': base_p,
                'optProfit': opt_p,
                'stations': {}
            }

            # 为每个站点存储优化结果
            for i, tid in enumerate(cat_tazs):
                cat_results['stations'][str(tid)] = {
                    1: float(S_opt_lab[0, i]),
                    2: float(S_opt_lab[1, i]),
                    3: float(S_opt_lab[2, i]),
                    4: float(S_opt_lab[3, i])
                }

            all_results[cat_name] = cat_results

            # 打印部分站点结果
            print(f"\nSample station results (first 5):")
            for i, tid in enumerate(cat_tazs[:5]):
                print(f"  TAZ {tid}: L1={S_opt_lab[0,i]:.4f}, L2={S_opt_lab[1,i]:.4f}, L3={S_opt_lab[2,i]:.4f}, L4={S_opt_lab[3,i]:.4f}")
        else:
            print(f"Optimization Failed: {res.message}")

    # 保存结果
    output_path = os.path.join(path, '../optimization-results-all.js')
    with open(output_path, 'w') as f:
        f.write("// 全量优化结果（所有站点）\n")
        f.write("window.OPTIMIZATION_RESULTS = ")
        json.dump(all_results, f, indent=2, ensure_ascii=False)
        f.write(";")

    print(f"\n\nResults saved to: {output_path}")

    # 打印汇总表
    print(f"\n{'='*80}")
    print(f"{'Category':<28} | {'Base Profit':<12} | {'Opt Profit':<12} | {'Increase':<10}")
    print("-" * 80)
    for cat_name, results in all_results.items():
        print(f"{cat_name:<28} | {results['baseProfit']:12.2f} | {results['optProfit']:12.2f} | {results['profitIncrease']:8.2f}%")


if __name__ == "__main__":
    run_optimization_all()

File: scripts/optimize_all_stations.js
// 全量优化脚本 - 为所有站点计算最优定价
const fs = require('fs');
const path = require('path');

// 加载数据
const dataPath = path.join(__dirname, '../data/');

// 读取 CSV 辅助函数
function readCSV(filename) {
    const lines = fs.readFileSync(path.join(dataPath, filename), 'utf-8').split('\n');
    const headers = lines[0].split(',');
    const data = {};
    headers.forEach((h, i) => data[h.trim()] = []);
    
    for (let i = 1; i < lines.length; i++) {
        if (!lines[i].trim()) continue;
        const vals = lines[i].split(',');
        headers.forEach((h, j) => {
            const v = vals[j]?.trim();
            data[h][i-1] = isNaN(v) ? v : parseFloat(v);
        });
    }
    return data;
}

// 读取 JS 变量文件
function loadJSVar(filename, varName) {
    const content = fs.readFileSync(path.join(dataPath, filename), 'utf-8');
    const match = content.match(new RegExp(`${varName}\\s*=\\s*(\\[[\\s\\S]*?\\])\\s*;`));
    if (match) {
        return JSON.parse(match[1]);
    }
    return null;
}

// 聚类时段映射（使用 poiRaw 小写）
const HOUR_LABELS = {
    'Lifestyle Services': {1: [10, 11, 14, 15, 16, 17, 18], 2: [7, 8, 9, 12, 13, 19, 20, 21], 3: [22, 23], 4: [0, 1, 2, 3, 4, 5, 6]},
    'Food and Beverage Services': {1: [14, 15, 16, 17, 18, 19], 2: [8, 9, 10, 20, 21, 22, 23], 3: [2, 3, 4, 5, 6, 7, 11, 12, 13], 4: [0, 1]},
    'Business and Residential': {1: [10, 11, 14, 15, 16, 17, 18], 2: [7, 8, 9, 12, 13, 19, 20], 3: [21, 22, 23], 4: [0, 1, 2, 3, 4, 5, 6]}
};

// 论文参数
const CATEGORIES = {
    'Lifestyle Services': {rho: -0.0724, beta1: -5.5852, beta2: 0.4636, const: 6.4749, b_t: 0.0038},
    'Business and Residential': {rho: -0.2358, beta1: -2.2836, beta2: 1.5628, const: 5.0514, b_t: 0.0045},
    'Food and Beverage Services': {rho: 0.0274, beta1: -6.0837, beta2: -0.2428, const: 6.6915, b_t: 0.0052}
};

// poiRaw 到英文类别的映射
const POI_RAW_TO_CAT = {
    'lifestyle services': 'Lifestyle Services',
    'food and beverage services': 'Food and Beverage Services',
    'business and residential': 'Business and Residential'
};

const T = 24, S_MIN = 0.20, S_MAX = 0.80, PHI_MAX = 50;
const C_BASE = 0.05, ETA = 0.2;

// 加载站点数据
const stationsData = loadJSVar('stations-data.js', 'TAZ_POINTS');

// 构建 TAZ -> 类别映射（使用 poiRaw）
const tazCatMap = {};
if (stationsData && Array.isArray(stationsData)) {
    stationsData.forEach(p => {
        if (p.tazid && p.poiRaw) {
            tazCatMap[p.tazid] = p.poiRaw.toLowerCase();
        }
    });
}

console.log('TAZ category mapping loaded:', Object.keys(tazCatMap).length);

// 加载价格和占用率数据
const infCsv = readCSV('inf.csv');
const ePriceCsv = readCSV('e_price.csv');
const sPriceCsv = readCSV('s_price.csv');
const occCsv = readCSV('occupancy.csv');

// 获取所有 TAZ
const allTazIds = [...new Set(infCsv.TAZID)];

console.log(`Total TAZs loaded: ${allTazIds.length}`);

// 获取每个 TAZ 的 charge_count
const tazChargeCount = {};
infCsv.TAZID.forEach((taz, i) => {
    if (!tazChargeCount[taz]) tazChargeCount[taz] = 0;
    tazChargeCount[taz] += infCsv.charge_count[i] || 1;
});

// 获取价格数据列名
const priceTazIds = Object.keys(ePriceCsv).filter(k => !isNaN(k) && k !== 'hour');

// 获取占用率数据
function getOccSeries(tazId) {
    const idx = priceTazIds.indexOf(String(tazId));
    if (idx >= 0 && occCsv[priceTazIds[idx]]) {
        return occCsv[priceTazIds[idx]].slice(0, T);
    }
    return Array(T).fill(0.35);
}

// 计算单站点需求
function calcDemand(sFee, ePrice, params, hour) {
    const P = Math.max(0.01, sFee + ePrice);
    const lnQ = params.const + params.b_t * (hour + 1) + params.beta1 * Math.log(P);
    return Math.exp(Math.max(-10, Math.min(10, lnQ)));
}

// 计算单站点利润
function calcStationProfit(sFees, ePrices, occSeries, params) {
    let totalProfit = 0;
    const occMax = Math.max(...occSeries, 0.01);
    
    for (let h = 0; h < T; h++) {
        const Q = calcDemand(sFees[h], ePrices[h], params, h);
        const cv = C_BASE * (1 + ETA * Math.pow(occSeries[h] / occMax, 2));
        totalProfit += Q * (sFees[h] - cv);
    }
    return totalProfit;
}

// 获取某小时对应的标签
function getHourLabel(hoursLabels, hour) {
    for (const [label, hours] of Object.entries(hoursLabels)) {
        if (hours.includes(hour)) return parseInt(label);
    }
    return 1;
}

// 获取标签对应的服务费
function getLabelFee(optFees, hoursLabels, hour) {
    const label = getHourLabel(hoursLabels, hour);
    return optFees[label] || 0.5;
}

// 单站点优化 - 使用网格搜索
function optimizeStation(ePrices, occSeries, params, hoursLabels) {
    // 网格搜索最优标签费用
    let bestFees = {1: 0.4, 2: 0.4, 3: 0.4, 4: 0.4};
    let bestProfit = -Infinity;
    
    const searchGrid = [0.20, 0.30, 0.40, 0.50, 0.60, 0.70, 0.80];
    
    // 简化：只优化4个标签级别
    for (const l1 of searchGrid) {
        for (const l2 of searchGrid) {
            for (const l3 of searchGrid) {
                for (const l4 of searchGrid) {
                    const fees = {1: l1, 2: l2, 3: l3, 4: l4};
                    const hourlyFees = Array(T).fill(0).map((_, h) => getLabelFee(fees, hoursLabels, h));
                    const profit = calcStationProfit(hourlyFees, ePrices, occSeries, params);
                    
                    if (profit > bestProfit) {
                        bestProfit = profit;
                        bestFees = {...fees};
                    }
                }
            }
        }
    }
    
    // 精细化搜索
    const fineGrid = [];
    for (let v = S_MIN; v <= S_MAX; v += 0.05) fineGrid.push(v);
    
    for (const l1 of fineGrid) {
        for (const l2 of fineGrid) {
            for (const l3 of fineGrid) {
                for (const l4 of fineGrid) {
                    const fees = {1: l1, 2: l2, 3: l3, 4: l4};
                    const hourlyFees = Array(T).fill(0).map((_, h) => getLabelFee(fees, hoursLabels, h));
                    const profit = calcStationProfit(hourlyFees, ePrices, occSeries, params);
                    
                    if (profit > bestProfit) {
                        bestProfit = profit;
                        bestFees = {...fees};
                    }
                }
            }
        }
    }
    
    return { fees: bestFees, profit: bestProfit };
}

// 主优化循环
const allResults = {};

for (const [catName, params] of Object.entries(CATEGORIES)) {
    console.log(`\n${'='.repeat(60)}`);
    console.log(`Processing: ${catName}`);
    console.log(`${'='.repeat(60)}`);
    
    // 获取该类别的所有站点（需要将 poiRaw 映射到英文类别名）
    const catTazs = allTazIds.filter(taz => {
        const poiRaw = tazCatMap[String(taz)] || tazCatMap[taz];
        const mappedCat = poiRaw ? POI_RAW_TO_CAT[poiRaw] : null;
        return mappedCat === catName && priceTazIds.includes(String(taz));
    });
    console.log(`Stations in category: ${catTazs.length}`);
    
    if (catTazs.length === 0) {
        console.log('No stations found!');
        continue;
    }
    
    const hoursLabels = HOUR_LABELS[catName];
    const catStations = {};
    let totalBaseProfit = 0;
    let totalOptProfit = 0;
    let successCount = 0;
    
    // 优化每个站点
    for (const tazId of catTazs) {
        const ePrices = ePriceCsv[String(tazId)]?.slice(0, T) || Array(T).fill(0.5);
        const sPrices = sPriceCsv[String(tazId)]?.slice(0, T) || Array(T).fill(0.5);
        const occSeries = getOccSeries(tazId);
        
        // 计算基础利润
        const baseProfit = calcStationProfit(sPrices, ePrices, occSeries, params);
        totalBaseProfit += baseProfit;
        
        // 优化
        try {
            const { fees, profit: optProfit } = optimizeStation(ePrices, occSeries, params, hoursLabels);
            
            catStations[tazId] = {
                1: parseFloat(fees[1].toFixed(4)),
                2: parseFloat(fees[2].toFixed(4)),
                3: parseFloat(fees[3].toFixed(4)),
                4: parseFloat(fees[4].toFixed(4))
            };
            
            totalOptProfit += optProfit;
            successCount++;
        } catch (err) {
            console.log(`  Error optimizing TAZ ${tazId}: ${err.message}`);
            catStations[tazId] = {1: 0.4, 2: 0.4, 3: 0.4, 4: 0.4};
        }
        
        if (catTazs.indexOf(tazId) % 50 === 0) {
            console.log(`  Progress: ${catTazs.indexOf(tazId) + 1}/${catTazs.length}`);
        }
    }
    
    const profitInc = totalBaseProfit !== 0 ? ((totalOptProfit - totalBaseProfit) / Math.abs(totalBaseProfit) * 100) : 0;
    
    allResults[catName] = {
        profitIncrease: parseFloat(profitInc.toFixed(2)),
        occIncrease: 0,
        baseProfit: parseFloat(totalBaseProfit.toFixed(2)),
        optProfit: parseFloat(totalOptProfit.toFixed(2)),
        stations: catStations
    };
    
    console.log(`\nResults:`);
    console.log(`  Base Profit: ${totalBaseProfit.toFixed(2)}`);
    console.log(`  Opt Profit: ${totalOptProfit.toFixed(2)}`);
    console.log(`  Increase: ${profitInc.toFixed(2)}%`);
    console.log(`  Success: ${successCount}/${catTazs.length}`);
    
    // 打印样本
    const sampleTazs = catTazs.slice(0, 3);
    console.log(`\nSample station results:`);
    sampleTazs.forEach(taz => {
        const s = catStations[taz];
        console.log(`  TAZ ${taz}: L1=${s[1]}, L2=${s[2]}, L3=${s[3]}, L4=${s[4]}`);
    });
}

// 保存结果
const outputPath = path.join(dataPath, 'optimization-results-all.js');
const output = `// 全量优化结果（所有站点）\nwindow.OPTIMIZATION_RESULTS = ${JSON.stringify(allResults, null, 2)};`;

fs.writeFileSync(outputPath, output, 'utf-8');

console.log(`\n${'='.repeat(60)}`);
console.log(`Results saved to: ${outputPath}`);
console.log(`${'='.repeat(60)}`);

// 汇总表
console.log(`\nSummary:`);
console.log('Category                  | Base Profit  | Opt Profit   | Increase');
console.log('------------------------------------------------------------------');
for (const [catName, results] of Object.entries(allResults)) {
    console.log(`${catName.padEnd(24)} | ${results.baseProfit.toFixed(2).padStart(12)} | ${results.optProfit.toFixed(2).padStart(12)} | ${results.profitIncrease.toFixed(2)}%`);
}
